### PR TITLE
ARROW-7427: [Python] Support SparseCSFTensor

### PR DESCRIPTION
**Pull request by rok**
This is to solve [ARROW-7427](https://issues.apache.org/jira/browse/ARROW-7427).

### CODE REVIEW

**Comment by rok** [comment]
[test_sparse_csf_tensor_numpy_roundtrip](https://github.com/apache/arrow/pull/6558/files#diff-80711f7667e9c624365c63c9112dbf2dR335) appears to be [failing on windows](https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/31326874/job/dmita0b6fxcoh5n4#L4055). I suspect it's a cython issue in either [`SparseCSFTensorToNdarray`](https://github.com/apache/arrow/pull/6558/files#diff-365362847f745a23bee254dbb46b3db9R361) or [`NdarraysToSparseCSFTensor`](https://github.com/apache/arrow/pull/6558/files#diff-365362847f745a23bee254dbb46b3db9R449). @pitrou could you take a quick look at this please?

**Comment by pitrou** [comment]
Should this be reviewed after #6521?

**Comment by rok** [comment]
> Should this be reviewed after #6521?

Perhaps that'd be best.

**Comment by rok** [comment]
I figured out the issue I think. I wasn't forcing the `indptr` and `indices` arrays to 64 bit integers which is fine on linux and mac but fails on windows where default appears to be 32 bit integers.

**Comment by rok** [comment]
Rebased on ARROW-7419 #6521.

**Comment by pitrou** [comment]
@rok Looks like you'll need to fix the conflicts now that https://github.com/apache/arrow/pull/6521 was merged (hopefully `git rebase` will do).

**Comment by rok** [comment]
@pitrou rebased :).

File: cpp/src/arrow/python/numpy_convert.cc [hunk 1]
```diff
@@ -384,6 +384,40 @@ Status SparseCSCMatrixToNdarray(const std::shared_ptr<SparseCSCMatrix>& sparse_t
...
+      *sparse_tensor, {sparse_tensor->non_zero_length(), 1}, base, result_data.ref()));
+
+  // Wrap indices
+  *out_indptr = PyList_New(0);
+  *out_indices = PyList_New(0);
+
+  for (const auto& indptr : sparse_index.indptr()) {
+    PyObject* item;
+    RETURN_NOT_OK(TensorToNdarray(indptr, base, &item));
+    if (PyList_Append(*out_indptr, item) < 0) {
```
Comment (pitrou) [inline cpp/src/arrow/python/numpy_convert.cc]: Need to `Py_DECREF(item)` after this, as `PyList_Append` doesn't steal its reference.

File: cpp/src/arrow/python/numpy_convert.cc [hunk 2]
```diff
@@ -384,6 +384,40 @@ Status SparseCSCMatrixToNdarray(const std::shared_ptr<SparseCSCMatrix>& sparse_t
...
+      RETURN_IF_PYERROR();
+    }
+  }
+  for (const auto& indices : sparse_index.indices()) {
+    PyObject* item;
+    RETURN_NOT_OK(TensorToNdarray(indices, base, &item));
+    if (PyList_Append(*out_indices, item) < 0) {
```
Comment (pitrou) [inline cpp/src/arrow/python/numpy_convert.cc]: Same here.

File: cpp/src/arrow/python/numpy_convert.cc [hunk 3]
```diff
@@ -384,6 +384,40 @@ Status SparseCSCMatrixToNdarray(const std::shared_ptr<SparseCSCMatrix>& sparse_t
...
+
+  // Wrap indices
+  *out_indptr = PyList_New(0);
+  *out_indices = PyList_New(0);
+
+  for (const auto& indptr : sparse_index.indptr()) {
+    PyObject* item;
+    RETURN_NOT_OK(TensorToNdarray(indptr, base, &item));
+    if (PyList_Append(*out_indptr, item) < 0) {
+      RETURN_IF_PYERROR();
```
Comment (pitrou) [inline cpp/src/arrow/python/numpy_convert.cc]: Also need to `Py_CLEAR(*out_indptr)` and `Py_CLEAR(*out_indices)` before erroring out. You could probably use OwnedRef instead, to make things simpler.

File: python/pyarrow/__init__.py [hunk 4]
```diff
@@ -72,6 +72,7 @@ def parse_git(root, **kwargs):
                          Array, Tensor,
                          array, chunked_array, record_batch, table,
                          SparseCOOTensor, SparseCSRMatrix, SparseCSCMatrix,
+                         SparseCSRMatrix, SparseCOOTensor, SparseCSFTensor,
```
Comment (pitrou) [inline python/pyarrow/__init__.py]: Did `SparseCSCMatrix` disappear?

File: python/pyarrow/tensor.pxi [hunk 5]
```diff
@@ -702,3 +705,156 @@ shape: {0.shape}""".format(self)
...
+        """
+        cdef shared_ptr[CSparseCSFTensor] csparse_tensor
+        cdef vector[int64_t] c_axis_order
+        cdef vector[int64_t] c_shape
+        cdef vector[c_string] c_dim_names
+
+        for x in shape:
+            c_shape.push_back(x)
+        if not axis_order:
+            axis_order = list(range(len(shape)))
```
Comment (pitrou) [inline python/pyarrow/tensor.pxi]: Isn't it better to use a heuristic based on dimension sizes here?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for doing this. A couple things stick out.

File: python/pyarrow/tensor.pxi (hunk 5 above)
Comment (rok) [reply python/pyarrow/tensor.pxi]: Good idea, changing to `argsort` heuristic.
Alternatively we can also just raise as `axis_order` is quite critical. I slightly prefer `argsort`.
I'm thinking to also add `SparseCSFTensor.axis_order` property.

File: python/pyarrow/__init__.py (hunk 4 above)
Comment (rok) [reply python/pyarrow/__init__.py]: It was doubled actually. Bad merge. Fixed.

File: cpp/src/arrow/python/numpy_convert.cc (hunk 1 above)
Comment (rok) [reply cpp/src/arrow/python/numpy_convert.cc]: Switching to `PyList_SET_ITEM`.

File: cpp/src/arrow/python/numpy_convert.cc (hunk 3 above)
Comment (rok) [reply cpp/src/arrow/python/numpy_convert.cc]: Made changes please check.

**Comment by rok** [comment]
CI seems unrelated.

File: python/pyarrow/tensor.pxi (hunk 5 above)
Comment (pitrou) [reply python/pyarrow/tensor.pxi]: Yes, I think `argsort` is more user-friendly.

**Review by pitrou — APPROVED** [review]
+1. Will merge if green.

File: cpp/src/arrow/python/numpy_convert.cc (hunk 3 above)
Comment (rok) [reply cpp/src/arrow/python/numpy_convert.cc]: Got it. Thanks!

**Pull request closed without merging**